### ISSUE TITLE
DraGAN

### ISSUE THREAD

**Issue by roddtalebi** [ezCGP/ezCGP#216]
Implement draGAN into loss.
The training occurs in codes/individual_definitions/individual_evaluate.py but the loss used in training is defined and added to a config dictionary.

Right now the loss is hard coded and set in codes/block_definitions/utilities/operators_simgan_train_config.py

If the dragan code is really long and ugly, I suggest putting it in it's own file in that same utilities folder or in ~/misc and then import it in and set it in that operators_simgan_train_config.py script

**Comment by roddtalebi** [comment]
all i did so far is make the branch and copy the code over. gotta look it over and see how best to implement

**Comment by roddtalebi** [comment]
TODO verify if we should be using 'xavier_init' if using dragan gradient penalty

**Comment by gabriel-q-wang** [comment]
Yep, the original dragan paper specified using xavier initialization for the weights in their code, but not their paper. Unsure of how big of an impact it has. https://github.com/kodalinaveen3/DRAGAN

**Comment by roddtalebi** [comment]
Also what should I set the 'penalty_constant' to be? How likely is it that something like that will change?

**Comment by gabriel-q-wang** [comment]
The penalty_constant is technically a hyperparameter than can be changed, but all the literature just sets it at 10 for some reason.

The local loss function we used was https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html cause of some internal speedup or something.

As for the actual "local" part of it, I was hoping we could discuss that a bit tomorrow as technically that should be another discriminator, and should hopefully be an optional block in the individual, mentioned in Issue#225

**Comment by roddtalebi** [comment]
Got the branch working with multiple generations...just need to verify somehow that if someone dies that it is for a good/right reason and not because the code sucks...that will not be a fun day of debugging
